gadget/ondisk.go

```diff
@@ -101,165 +55,103 @@ type OnDiskVolume struct {
-}
-
-func fromSfdiskPartitionType(st string, sfdiskLabel string) (string, error) {
-	switch sfdiskLabel {
-	case "dos":
-		// sometimes sfdisk reports what is "0C" in gadget.yaml as "c",
-		// normalize the values
-		v, err := strconv.ParseUint(st, 16, 8)
-		if err != nil {
-			return "", fmt.Errorf("cannot convert MBR partition type %q", st)
-		}
-		return fmt.Sprintf("%02X", v), nil
-	case "gpt":
-		return st, nil
-	default:
-		return "", fmt.Errorf("unsupported partitioning schema %q", sfdiskLabel)
-	}
-}
-
-func blockdevSizeCmd(cmd, devpath string) (quantity.Size, error) {
-	out, err := exec.Command("blockdev", cmd, devpath).CombinedOutput()
-	if err != nil {
-		return 0, osutil.OutputErr(out, err)
-	}
-	nospace := strings.TrimSpace(string(out))
-	sz, err := strconv.Atoi(nospace)
-	if err != nil {
-		return 0, fmt.Errorf("cannot parse blockdev %s result size %q: %v", cmd, nospace, err)
-	}
-	return quantity.Size(sz), nil
-}
 
-func blockDeviceSizeInSectors(devpath string) (quantity.Size, error) {
-	// the size is always reported in 512-byte sectors, even if the device does
-	// not have a physical sector size of 512
-	// XXX: consider using /sys/block/<dev>/size directly
-	return blockdevSizeCmd("--getsz", devpath)
+	return OnDiskVolumeFromDisk(disk)
 }
 
-func blockDeviceSectorSize(devpath string) (quantity.Size, error) {
-	// the size is reported in raw bytes
-	sz, err := blockdevSizeCmd("--getss", devpath)
+func OnDiskVolumeFromDisk(disk disks.Disk) (*OnDiskVolume, error) {
+	parts, err := disk.Partitions()
 	if err != nil {
-		return 0, err
-	}
-
-	// ensure that the sector size is a multiple of 512, since we rely on that
-	// when we calculate the size in sectors, as blockdev --getsz always returns
-	// the size in 512-byte sectors
-	if sz%512 != 0 {
-		return 0, fmt.Errorf("cannot calculate structure size: sector size (%s) is not a multiple of 512", sz.String())
-	}
-	if sz == 0 {
-		// extra paranoia
-		return 0, fmt.Errorf("internal error: sector size returned as 0")
+		return nil, err
 	}
-	return sz, nil
-}
 
-// onDiskVolumeFromPartitionTable takes an sfdisk dump partition table and returns
-// the partitioning information as an on-disk volume.
-func onDiskVolumeFromPartitionTable(ptable sfdiskPartitionTable) (*OnDiskVolume, error) {
-	if ptable.Unit != "sectors" {
-		return nil, fmt.Errorf("cannot position partitions: unknown unit %q", ptable.Unit)
-	}
+	structure := make([]VolumeStructure, len(parts))
+	ds := make([]OnDiskStructure, len(parts))
 
-	structure := make([]VolumeStructure, len(ptable.Partitions))
-	ds := make([]OnDiskStructure, len(ptable.Partitions))
+	for _, p := range parts {
+		// use the index of the structure on the disk rather than the order in
+		// which we iterate over the list of partitions, since the order of the
+		// partitions is returned "last seen first" which matches the behavior
+		// of udev when picking partitions with the same filesystem label and
+		// populating /dev/disk/by-label/ and friends
```

Comment:
nitpick: add a period here, otherwise one might continue to read the line below thinking that it's part of the same sentence.

Reply:
done